File: sample/kotlin/08_langtons_ant.kt
import kotlin.math.*


// 08: Sample that outputs Langton's Ant trajectories as a GIF.

fun capture(grid: MutableList<Any?>, w: Long, h: Long): MutableList<Any?> {
    var frame: MutableList<Any?> = __pytra_as_list(__pytra_bytearray((__pytra_int(w) * __pytra_int(h))))
    val __step_0 = __pytra_int(1L)
    var y = __pytra_int(0L)
    while ((__step_0 >= 0L && y < __pytra_int(h)) || (__step_0 < 0L && y > __pytra_int(h))) {
        var row_base: Long = __pytra_int((__pytra_int(y) * __pytra_int(w)))
        val __step_1 = __pytra_int(1L)
        var x = __pytra_int(0L)
        while ((__step_1 >= 0L && x < __pytra_int(w)) || (__step_1 < 0L && x > __pytra_int(w))) {
            __pytra_set_index(frame, (__pytra_int(row_base) + __pytra_int(x)), __pytra_ifexp((__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)) != 0L), 255L, 0L))
            x += __step_1
        }
        y += __step_0
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

fun run_08_langtons_ant() {
    var w: Long = __pytra_int(420L)
    var h: Long = __pytra_int(420L)
    var out_path: String = __pytra_str("sample/out/08_langtons_ant.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var grid: MutableList<Any?> = __pytra_as_list(run { val __out = mutableListOf<Any?>(); val __step = __pytra_int(1L); var __lc_i = __pytra_int(0L); while ((__step >= 0L && __lc_i < __pytra_int(h)) || (__step < 0L && __lc_i > __pytra_int(h))) { __out.add(__pytra_list_repeat(0L, w)); __lc_i += __step }; __out })
    var x: Long = __pytra_int((__pytra_int(__pytra_int(w) / __pytra_int(2L))))
    var y: Long = __pytra_int((__pytra_int(__pytra_int(h) / __pytra_int(2L))))
    var d: Long = __pytra_int(0L)
    var steps_total: Long = __pytra_int(600000L)
    var capture_every: Long = __pytra_int(3000L)
    var frames: MutableList<Any?> = __pytra_as_list(mutableListOf<Any?>())
    val __step_0 = __pytra_int(1L)
    var i = __pytra_int(0L)
    while ((__step_0 >= 0L && i < __pytra_int(steps_total)) || (__step_0 < 0L && i > __pytra_int(steps_total))) {
        if ((__pytra_int(__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x))) == __pytra_int(0L))) {
            d = __pytra_int((__pytra_int((__pytra_int(d) + __pytra_int(1L))) % __pytra_int(4L)))
            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, 1L)
        } else {
            d = __pytra_int((__pytra_int((__pytra_int(d) + __pytra_int(3L))) % __pytra_int(4L)))
            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, 0L)
        }
        if ((__pytra_int(d) == __pytra_int(0L))) {
            y = __pytra_int((__pytra_int((__pytra_int((__pytra_int(y) - __pytra_int(1L))) + __pytra_int(h))) % __pytra_int(h)))
        } else {
            if ((__pytra_int(d) == __pytra_int(1L))) {
                x = __pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(1L))) % __pytra_int(w)))
            } else {
                if ((__pytra_int(d) == __pytra_int(2L))) {
                    y = __pytra_int((__pytra_int((__pytra_int(y) + __pytra_int(1L))) % __pytra_int(h)))
                } else {
                    x = __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) - __pytra_int(1L))) + __pytra_int(w))) % __pytra_int(w)))
                }
            }
        }
        if ((__pytra_int((__pytra_int(i) % __pytra_int(capture_every))) == __pytra_int(0L))) {
            frames = __pytra_as_list(frames); frames.add(capture(grid, w, h))
        }
        i += __step_0
    }
    __pytra_noop(out_path, w, h, frames, mutableListOf<Any?>())
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", __pytra_len(frames))
    __pytra_print("elapsed_sec:", elapsed)
}

fun main(args: Array<String>) {
    run_08_langtons_ant()
}

File: sample/swift/08_langtons_ant.swift
import Foundation


// 08: Sample that outputs Langton's Ant trajectories as a GIF.

func capture(grid: [Any], w: Int64, h: Int64) -> [Any] {
    var frame: [Any] = __pytra_as_list(__pytra_bytearray((w * h)))
    var y = __pytra_int(Int64(0))
    while (y < __pytra_int(h)) {
        var row_base: Int64 = (y * w)
        var x = __pytra_int(Int64(0))
        while (x < __pytra_int(w)) {
            __pytra_setIndex(frame, (row_base + x), __pytra_ifexp((__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)) != 0), Int64(255), Int64(0)))
            x += 1
        }
        y += 1
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

func run_08_langtons_ant() {
    var w: Int64 = Int64(420)
    var h: Int64 = Int64(420)
    var out_path: String = "sample/out/08_langtons_ant.gif"
    var start: Double = __pytra_perf_counter()
    var grid: [Any] = __pytra_as_list(({ () -> [Any] in var __out: [Any] = []; let __step = __pytra_int(Int64(1)); var __lc_i = __pytra_int(Int64(0)); while ((__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h))) { __out.append(__pytra_list_repeat(Int64(0), w)); __lc_i += __step }; return __out })())
    var x: Int64 = (w / Int64(2))
    var y: Int64 = (h / Int64(2))
    var d: Int64 = Int64(0)
    var steps_total: Int64 = Int64(600000)
    var capture_every: Int64 = Int64(3000)
    var frames: [Any] = __pytra_as_list([])
    var i = __pytra_int(Int64(0))
    while (i < __pytra_int(steps_total)) {
        if (__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)) == __pytra_int(Int64(0))) {
            d = ((d + Int64(1)) % Int64(4))
            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(1))
        } else {
            d = ((d + Int64(3)) % Int64(4))
            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(0))
        }
        if (__pytra_int(d) == __pytra_int(Int64(0))) {
            y = (((y - Int64(1)) + h) % h)
        } else {
            if (__pytra_int(d) == __pytra_int(Int64(1))) {
                x = ((x + Int64(1)) % w)
            } else {
                if (__pytra_int(d) == __pytra_int(Int64(2))) {
                    y = ((y + Int64(1)) % h)
                } else {
                    x = (((x - Int64(1)) + w) % w)
                }
            }
        }
        if (__pytra_int(i % capture_every) == __pytra_int(Int64(0))) {
            frames.append(capture(grid, w, h))
        }
        i += 1
    }
    __pytra_noop(out_path, w, h, frames, [])
    var elapsed: Double = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", __pytra_len(frames))
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_08_langtons_ant()
    }
}
